subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,215 @@
+
+static void audio_interface_config(struct usb_desc_header *head,
+				      u8_t bInterfaceNumber)
+{
+	ARG_UNUSED(head);
+	LOG_WRN("CFG");
+}
+
+void audio_cb_usb_status(struct usb_cfg_data *cfg,
+			      enum usb_dc_status_code cb_status,
+			      const u8_t *param)
+{
+	// LOG_WRN("CB %08x", (int)cb_status);
+}
+
+/**
+ * @brief Handler called for Class requests not handled by the USB stack.
+ *
+ * @param pSetup    Information about the request to execute.
+ * @param len       Size of the buffer.
+ * @param data      Buffer containing the request result.
+ *
+ * @return  0 on success, negative errno code on fail.
+ */
+int audio_class_handle_req(struct usb_setup_packet *pSetup,
+			   s32_t *len, u8_t **data)
+{
+	uint32_t wIndex;
+	LOG_WRN("Req %d",(u32_t)pSetup->bRequest);
+
+	struct usb_audio_dev_data_t *dev_data;
+	struct usb_dev_data *common;
+
+	LOG_DBG("Called audio class handle req");
+
+	wIndex = sys_le16_to_cpu(pSetup->wIndex) & 0xFF;
+	common = usb_get_dev_data_by_iface(&usb_audio_data_devlist,
+			wIndex);
+	if (common == NULL) {
+		LOG_DBG("Device data not found for interface %u", wIndex);
+		return -ENODEV;
+	}
+
+	dev_data = CONTAINER_OF(common, struct usb_audio_dev_data_t, common);
+
+	switch (pSetup->bRequest)
+	{
+	case 0x81:
+		dev_data->mixerCurrent = 1;
+		*data = &(dev_data->mixerCurrent);
+		*len = 1;
+		LOG_WRN("GET_CUR");
+		break;
+
+	case 0x1:
+		*data = &(dev_data->mixerCurrent);
+		*len = 1;
+		LOG_WRN("SET_CUR");
+		break;
+
+	default:
+		LOG_DBG("Audio request 0x%x, value 0x%x",
+				pSetup->bRequest, pSetup->wValue);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct usb_audio_device_api {
+	void (*init)(void);
+} usb_audio_api;
+
+static int usb_audio_device_init(struct device *dev)
+{
+	struct usb_audio_dev_data_t * const dev_data = (void*) dev->driver_data;
+	int ret = 0;
+
+	dev_data->common.dev = dev;
+	sys_slist_append(&usb_audio_data_devlist, &dev_data->common.node);
+
+	LOG_DBG("Device dev %p dev_data %p cfg %p added to devlist %p",
+		dev, dev_data, dev->config->config_info, &usb_audio_data_devlist);
+
+	return ret;
+}
+
+#define DEFINE_AUDIO_DEVICE(x)						\
+	DEVICE_AND_API_INIT(usb_audio_device_##x,			\
+			    CONFIG_USB_AUDIO_DEVICE_NAME_##x,		\
+			    &usb_audio_device_init,			\
+			    &audio_dev_data_##x,			\
+			    &audio_config_##x, APPLICATION,		\
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+			    &usb_audio_api)
+
+
+void recv_callback(u8_t ep, enum usb_dc_ep_cb_status_code status)
+{
+	int ret;
+	int ret_bytes;
+	struct usb_audio_dev_data_t *dev_data;
+	struct usb_dev_data *common;
+
+	ARG_UNUSED(status);
+
+	common = usb_get_dev_data_by_ep(&usb_audio_data_devlist, ep);
+	if (common == NULL) {
+		return;
+	}
+
+	dev_data = CONTAINER_OF(common, struct usb_audio_dev_data_t, common);
+
+	ret = usb_read(ep, dev_data->rx_buf, AUDIO_RX_BUFFER_SIZE, &ret_bytes);
+	if (ret) {
+		LOG_ERR("ret=%d ", ret);
+	}
+}
+
+DEFINE_AUDIO_HEADPHONES_DESCRIPTOR(0, 0x8);
```

Comment:
This should be first unused iso capable endpoint instead of magic number. Something like `AUTO_EP_IN/OUT` that would also check if the endpoint can handle isochronous transfers.